applications/tari_validator_node/src/p2p/rpc/service_impl.rs

```diff
@@ -119,4 +143,90 @@ where TPeerProvider: PeerProvider + Clone + Send + Sync + 'static
 
         Ok(Streaming::new(rx))
     }
+
+    async fn get_vn_state_inventory(
+        &self,
+        request: Request<GetVnStateInventoryRequest>,
+    ) -> Result<Response<GetVnStateInventoryResponse>, RpcStatus> {
+        let request = request.into_message();
+        let start_shard_id = request
+            .start_shard_id
+            .and_then(|s| ShardId::try_from(s).ok())
+            .ok_or_else(|| RpcStatus::bad_request("Invalid gRPC request: start_shard_id not provided"))?;
+        let end_shard_id = request
+            .end_shard_id
+            .and_then(|s| ShardId::try_from(s).ok())
+            .ok_or_else(|| RpcStatus::bad_request("Invalid gRPC request: end_shard_id not provided"))?;
+
+        let store_tx = self
+            .shard_state_store
+            .create_tx()
+            .map_err(RpcStatus::log_internal_error(LOG_TARGET))?;
+        let inventory = store_tx
+            .get_state_inventory(start_shard_id, end_shard_id)
+            .map_err(RpcStatus::log_internal_error(LOG_TARGET))?;
+
+        Ok(Response::new(GetVnStateInventoryResponse {
+            inventory: inventory.into_iter().map(|item| item.into()).collect(),
+        }))
+    }
+
+    async fn vn_state_sync(
+        &self,
+        request: Request<VnStateSyncRequest>,
+    ) -> Result<Streaming<VnStateSyncResponse>, RpcStatus> {
+        let (tx, rx) = mpsc::channel(100);
+        let msg = request.into_message();
+
+        let missing_shard_ids = msg
+            .missing_shard_state_ids
+            .iter()
+            .map(|s| {
+                ShardId::try_from(s.bytes.as_slice())
+                    .expect("Invalid gRPC request: failed to parse shard id's request data")
+            })
+            .collect::<Vec<ShardId>>();
+
+        if missing_shard_ids.is_empty() {
```

Comment:
Edge case here, and if the network is running very likely that won't ever be empty after a small period. 
But when the network just started, and the first nodes boot up, they will fail to sync as there will be no shards available.